lib/models/unfurl.js

```diff
@@ -0,0 +1,226 @@
+const githubUrl = require('../github-url');
+
+class UnsupportedResource extends Error {
+  constructor(url) {
+    super(url);
+    this.name = 'UnsupportedResource';
+  }
+}
+
+/* eslint-disable global-require */
+const resources = {
+  account: require('../unfurls/account'),
+  blob: require('../unfurls/blob'),
+  comment: require('../unfurls/comment'),
+  issue: require('../unfurls/issue'),
+  pull: require('../unfurls/pull'),
+  repo: require('../unfurls/repo'),
+};
+
+module.exports = (sequelize, DataTypes) => {
+  const { SlackWorkspace, SlackUser, GitHubUser } = sequelize.models;
+  const Unfurl = sequelize.define('Unfurl', {
+    url: {
+      type: DataTypes.STRING,
+      allowNull: false,
+    },
+    channelSlackId: {
+      type: DataTypes.STRING,
+      allowNull: false,
+    },
+    githubType: {
+      type: DataTypes.STRING,
+      allowNull: true,
+    },
+    isCondensed: {
+      type: DataTypes.BOOLEAN,
+      allowNull: false,
+    },
+    isPublic: {
+      type: DataTypes.BOOLEAN,
+      allowNull: false,
+    },
+    isDelivered: {
+      type: DataTypes.BOOLEAN,
+      allowNull: false,
+    },
+    slackMessageTimestamp: {
+      type: DataTypes.STRING,
+      allowNull: false,
+    },
+  });
+
+  Unfurl.associate = (models) => {
+    Unfurl.belongsTo(models.SlackWorkspace, {
+      foreignKey: 'slackWorkspaceId',
+      allowNull: false,
+    });
+    Unfurl.belongsTo(models.SlackUser, {
+      foreignKey: 'slackUserId',
+      allowNull: true,
+    });
+  };
+
+
+  Object.assign(Unfurl.prototype, {
+    async removeUnfurlEligibility() {
+      const slackWorkspace = await SlackWorkspace.findById(this.slackWorkspaceId);
+      const key = `${slackWorkspace.slackId}-${this.channelSlackId}-${this.url}`;
+      return cache.set(key, true);
+    },
+
+    async deliver() {
+      const slackUser = await SlackUser.findById(this.slackUserId);
+      const github = await this.constructor.getGitHubClient(this.channelSlackId, slackUser.slackId);
+      const {
+        attachment,
+        githubType,
+      } = await this.constructor.getAttachment(github, this.url, this.isCondensed, this.isPublic);
+      const unfurls = {
+        [this.url]: attachment,
+      };
+      logger.debug(unfurls, 'Unfurling links');
+      const slackWorkspace = await SlackWorkspace.findById(this.slackWorkspaceId);
+      const slackRes = await slackWorkspace.client.chat.unfurl({
+        ts: this.slackMessageTimestamp,
+        channel: this.channelSlackId,
+        unfurls,
+      });
+      logger.trace(slackRes, 'Unfurl complete');
+      await this.removeUnfurlEligibility();
+      this.githubType = githubType;
+      this.isDelivered = true;
+      await this.save();
+    },
+  });
+
+  Object.assign(Unfurl, {
+    async getGitHubClient(channelId, userSlackId) {
+      const github = new GitHubApi({ logger });
+
+      if (!process.env.ALLOWED_CHANNELS || !process.env.ALLOWED_CHANNELS.split(',').includes(channelId)) {
```

Comment:
> `process.env.ALLOWED_CHANNELS.split(',')`

This would be good to do once at the top of the file and assign to a constant to avoid having to split the string on every request.

Reply:
Just tried implementing this. It works fine in `unfurls/index.js`, but not in `models/unfurl.js` because that file is only required once. Probably fine for staging/production but it breaks testing, so I'm just going to stick to the existing implementation for now.